File: collatz/collatz.cpp
#include <iostream>
#include <chrono>
#include <cstdint>

std::uint64_t collatz(std::uint64_t n) {
    std::uint64_t value = n;
    std::uint64_t iterations = 0;
    while (value > 1) {
        if (value % 2 == 0) {
            value /= 2;
        } else {
            value = value * 3 + 1;
        }
        ++iterations;
    }
    return iterations;
}

int main() {
    // Start timer
    auto start = std::chrono::high_resolution_clock::now();

    std::uint64_t start_from = 2;
    std::uint64_t end_at    = 10000002;
    std::uint64_t total_iterations = 0;

    for (std::uint64_t i = start_from; i < end_at; ++i) {
        if (collatz(i) == 0) {
            break;
        }
        ++total_iterations;
    }

    // Stop timer
    auto end = std::chrono::high_resolution_clock::now();
    // Measure elapsed time as a floating-point number of seconds
    std::chrono::duration<double> elapsed = end - start;

    std::cout << "Total iterations: " << total_iterations << " ";
    std::cout << " Time elapsed: "
              << elapsed.count() << " s \n";
    return 0;
}

File: collatz/collatz.go
package main

import (
    "time"
    "fmt"
)

func collatz(n uint64) uint64 {
    collatzNumber := n
    var iterations uint64 = 0

    for collatzNumber > 1 {
        if collatzNumber % 2 == 0 {
            collatzNumber /= 2
        } else {
            collatzNumber = collatzNumber * 3 + 1
        }
        iterations++;
    }
    return uint64(iterations)
}

func main() {
    now := time.Now()
    var startFrom uint64 = 2
    var endAt uint64 = 10000002
    var total_iterations uint64 = 0;
    for i := startFrom; i < endAt; i++ {
        total_iterations++
        if collatz(i) == 0 {
            break
        }
    }
    fmt.Printf("Total iterations: %v ", total_iterations)
    fmt.Printf(" Time elapsed: %v \n", time.Since(now))
}
